packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts

```diff
@@ -1231,29 +1233,80 @@ class TcbIfOp extends TcbOp {
+    // for the case where the expression has an alias _and_ because we need the processed
+    // expression when generating the guard for the body.
+    const expressionScope = Scope.forNodes(this.tcb, this.scope, branch, [], null);
+    expressionScope.render().forEach(stmt => this.scope.addStatement(stmt));
 
-    if (branch.expressionAlias === null) {
-      branchParentScope = this.scope;
-    } else {
-      // If the expression is aliased, we create a scope with a variable containing the expression.
-      // Further down we'll use the variable instead of the expression itself in the `if` statement.
-      // This allows for the type of the alias to be narrowed.
-      branchParentScope = Scope.forNodes(this.tcb, this.scope, branch, [], null);
-      branchParentScope.render().forEach(stmt => this.scope.addStatement(stmt));
-    }
-
-    const branchScope = Scope.forNodes(this.tcb, branchParentScope, null, branch.children, null);
     const expression = branch.expressionAlias === null ?
-        tcbExpression(branch.expression, this.tcb, branchScope) :
-        branchScope.resolve(branch.expressionAlias);
+        tcbExpression(branch.expression, this.tcb, expressionScope) :
+        expressionScope.resolve(branch.expressionAlias);
+    this.branchExpressions.set(branch, expression);
+
+    const bodyScope = Scope.forNodes(
+        this.tcb, expressionScope, null, branch.children,
+        this.generateBranchGuard(index, expression));
 
     return ts.factory.createIfStatement(
-        expression, ts.factory.createBlock(branchScope.render()), this.generateBranch(index + 1));
+        expression, ts.factory.createBlock(bodyScope.render()), this.generateBranch(index + 1));
+  }
+
+  private generateBranchGuard(index: number, ownExpression: ts.Expression|null): ts.Expression
+      |null {
+    let prevExpressions: ts.Expression|null = null;
+
+    // Since event listeners are inside callbacks, type narrowing doesn't apply to them anymore.
+    // To recreate the behavior, we generate an expression that negates all the values of the
+    // branches _before_ the current one, and then we add the current branch's expression on top.
+    // For example `@if (expr === 1) {} @else if (expr === 2) {} @else if (expr === 3)`, the guard
+    // for the last expression will be `!(expr === 1) && !(expr === 2) && expr === 3`.
+    for (let i = 0; i < index; i++) {
+      const branch = this.block.branches[i];
+
+      // This shouldn't happen, but we assert it anyway because it's likely
+      // it'll break silently if we make changes to it later.
+      if (!this.branchExpressions.has(branch)) {
+        throw new Error('Could not determine expression of branch at index ' + i);
+      }
+
+      const expression = this.branchExpressions.get(branch) as ts.Expression | null;
+
+      // Skip over branches without an expression.
+      if (expression !== null) {
```

Comment:
nit: compiler generally prefers early exits to reduce nesting

```suggestion
      if (expression === null) {
        continue;
      }
```